### DISCUSSION TITLE
Devtools 404 error

### DISCUSSION THREAD

**Discussion by JackVanson** [Laravel-Backpack/community-forum#1194] [Q&A (Help)] [answered]
Hi guys I am using Multi Tenancy for Laravel with Backpack and I have installed DevTools. When I try to open the devtools url. "/admin/devtools/model" I get the backpack 404 error page. Multi Tenancy for Laravel has two routes files one for the central site/database and one for the tenant site/database. I have copied the devtool routes from vendor\backpack\devtools\routes\devtools.php in to both of these routes file and still no luck. 

Can any one help?

**Comment by JackVanson** [comment]
I seem found a hack but i now need to make a it work. if the devtools.php routes file. "vendor\backpack\devtools\routes\devtools.php"

```
<?php
Route::group([
    'prefix' => config('backpack.base.route_prefix', 'admin').'/devtools',
    'middleware' => ['web', config('backpack.base.middleware_key', 'admin')],
    'namespace' => 'Backpack\DevTools\Http\Controllers',
    'name' => 'devtools.',
], function () {
    Route::get('/', function () {
        return redirect(config('backpack.base.route_prefix', 'admin').'/devtools/model');
    });

    Route::crud('model', 'ModelCrudController');
    Route::crud('migration', 'MigrationCrudController');
    Route::crud('operation', 'OperationCrudController');

    Route::get('dump-autoload', 'DumpAutoload');
});
```


if i remove the middleware it works. Not a great fix. Anyone else have a better option?

```
<?php
Route::group([
    'prefix' => config('backpack.base.route_prefix', 'admin').'/devtools',
    'namespace' => 'Backpack\DevTools\Http\Controllers',
    'name' => 'devtools.',
], function () {
    Route::get('/', function () {
        return redirect(config('backpack.base.route_prefix', 'admin').'/devtools/model');
    });

    Route::crud('model', 'ModelCrudController');
    Route::crud('migration', 'MigrationCrudController');
    Route::crud('operation', 'OperationCrudController');

    Route::get('dump-autoload', 'DumpAutoload');
});
```

**Comment by JackVanson** [comment]
I also get a 404 error with 

\vendor\livewire\livewire\src\Mechanisms\HandleRequests\HandleRequests.php

```
    function boot()
    {
        // Only set it if another provider hasn't already set it....
        if (! $this->updateRoute) {
            app($this::class)->setUpdateRoute(function ($handle) {
                return Route::post('/livewire/update', $handle)->middleware('web');
            });
        }

        $this->skipRequestPayloadTamperingMiddleware();
    }
```

If I remove the ->middleware('web') no 404 error.

**Comment by jcastroa87** [comment]
Hello @JackVanson 

To be honest, I never try to use Devtools on Multi Tenancy project, can you share exactly which package are you using.

In other hand, I think this kind of package are very specific about how routes need to be handled, so I think not all the package will be fully compatible without customize.

After you copy routes, did you try with:
```
php artisan cache:clear
```

Let me know more information to check if we can help you.

Cheers.

**Reply by JackVanson** [reply]
Morning @jcastroa87 I am using https://tenancyforlaravel.com/docs/v3/introduction v3 and BackPack v6.

I did clear all the cache after copying the devtool routes. No luck. Even after making the above changes i was still hit with some errors. E.g. where when i tried to save a new model i got a 403 on POST admin/devtools/model.

For now I have reverted my changes and will just the Laravel's commands to develop with. As i have paid for Devtools it would be great to get it working. However if i can't I will see if i can get a refund and work the old fashioned way hahah.

Thank you for your help!

Jack

**Reply by JackVanson** [reply]
So I had the same issue with the permissions manager but i have been able to fix that. You can override the permission manager routes file. 

From 
```
Route::group([
    'namespace'  => 'Backpack\PermissionManager\app\Http\Controllers',
    'prefix'     => config('backpack.base.route_prefix', 'admin'),
    'middleware' => ['web', backpack_middleware()],
], function () {
    foreach (config('tenancy.central_domains') as $domain) {
        Route::domain($domain)->group(function () {
            Route::crud('permission', 'PermissionCrudController');
            Route::crud('role', 'RoleCrudController');
            Route::crud('user', 'UserCrudController');
        });
    }
});
```

To

```
Route::group([
    'namespace'  => 'Backpack\PermissionManager\app\Http\Controllers',
     'prefix' => config('backpack.base.route_prefix', 'admin'),
    'middleware' => array_merge(
        (array) config('backpack.base.web_middleware', 'web'),
        (array) config('backpack.base.middleware_key', 'admin'),
    ),
], function () {
    foreach (config('tenancy.central_domains') as $domain) {
        Route::domain($domain)->group(function () {
            Route::crud('permission', 'PermissionCrudController');
            Route::crud('role', 'RoleCrudController');
            Route::crud('user', 'UserCrudController');
        });
    }
});
```

So my question is how can i override the devTools route file. so i can apply the same fix.

**Reply by JackVanson** [reply, marked as answer]
I've worked it out. Keeping a note on here just in case someone else has the same issue. I created my own DevToolsRouteServiceProvider, making sure i use php artisan to create it so it was registered in composer.json then I just override the routes in the boot. like so. 

```
    public function boot(): void
    {
        Route::group([
            'prefix' => config('backpack.base.route_prefix', 'admin').'/devtools',
            'middleware' => array_merge(
                (array) config('backpack.base.web_middleware', 'web'),
                (array) config('backpack.base.middleware_key', 'admin'),
            ),
            'namespace' => 'Backpack\DevTools\Http\Controllers',
            'name' => 'devtools.',
        ], function () {
            Route::get('/', function () {
                return redirect(config('backpack.base.route_prefix', 'admin').'/devtools/model');
            });
        
            Route::crud('model', 'ModelCrudController');
            Route::crud('migration', 'MigrationCrudController');
            Route::crud('operation', 'OperationCrudController');
        
            Route::get('dump-autoload', 'DumpAutoload');
        });
    }
```

**Comment by jcastroa87** [comment]
I'm glad you find the solution, and thanks for sharing with our community.

I will close this issue, but please feel free to re-open or create a new one if needed.

Cheers.
